File: app/fbcnms-projects/inventory/app/common/WorkOrder.js
/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

import type {AddWorkOrderTypeInput} from '../mutations/__generated__/AddWorkOrderTypeMutation.graphql';
import type {CheckListCategoryExpandingPanel_list} from '../components/checklist/checkListCategory/__generated__/CheckListCategoryExpandingPanel_list.graphql';
import type {ChecklistCategoryDefinition} from '../components/checklist/ChecklistCategoriesMutateState';
import type {Equipment, Link} from './Equipment';
import type {FileAttachmentType} from './FileAttachment.js';
import type {ImageAttachmentType} from './ImageAttachment.js';
import type {Location} from './Location';
import type {NamedNode} from './EntUtils';
import type {PriorityType, StatusType} from './FilterTypes';
import type {Property} from './Property';
import type {PropertyType} from './PropertyType';
import type {ShortUser} from './EntUtils';
import type {WorkOrderTemplateNodesQuery} from './__generated__/WorkOrderTemplateNodesQuery.graphql';

import {convertPropertyTypeToMutationInput} from './PropertyType';
import {graphql} from 'relay-runtime';
import {isTempId} from './EntUtils';
import {useLazyLoadQuery} from 'react-relay/hooks';

export type WorkOrderType = {
  id: string,
  name: string,
  description: ?string,
  propertyTypes: Array<PropertyType>,
  numberOfWorkOrders: number,
  checklistCategoryDefinitions: Array<ChecklistCategoryDefinition>,
};

export type WorkOrder = {
  id: string,
  workOrderType: ?WorkOrderType,
  workOrderTypeId: ?string,
  name: string,
  description: ?string,
  location: ?Location,
  locationId: ?string,
  owner: ShortUser,
  creationDate: string,
  installDate: ?string,
  status: StatusType,
  priority: PriorityType,
  equipmentToAdd: Array<Equipment>,
  equipmentToRemove: Array<Equipment>,
  linksToAdd: Array<Link>,
  linksToRemove: Array<Link>,
  images: Array<ImageAttachmentType>,
  files: Array<FileAttachmentType>,
  assignedTo: ?ShortUser,
  properties: Array<Property>,
  project: ?{
    id: string,
  },
  checkListCategories: ?CheckListCategoryExpandingPanel_list,
};

export type WorkOrderIdentifier = {
  +id: string,
  +name: string,
};

export type FutureState = 'INSTALL' | 'REMOVE';

export const FutureStateValues = [
  {
    key: 'install',
    value: 'INSTALL',
    label: 'Install',
  },
  {
    key: 'remove',
    value: 'REMOVE',
    label: 'Remove',
  },
];

export const convertWorkOrderTypeToMutationInput = (
  workOrderType: WorkOrderType,
): AddWorkOrderTypeInput => {
  return {
    name: workOrderType.name,
    description: workOrderType.description,
    properties: convertPropertyTypeToMutationInput(workOrderType.propertyTypes),
    checkListCategories: workOrderType.checklistCategoryDefinitions
      .slice()
      .map(categoryDef => ({
        ...categoryDef,
        id: isTempId(categoryDef.id) ? undefined : categoryDef.id,
        checkList: categoryDef.checkList.slice().map(item => ({
          ...item,
          id: isTempId(item.id) ? undefined : item.id,
        })),
      })),
  };
};

const workOrderTemplateNodesQuery = graphql`
  query WorkOrderTemplateNodesQuery {
    workOrderTypes {
      edges {
        node {
          id
          name
        }
      }
    }
  }
`;

export type WorkOrderTemplateNode = $Exact<NamedNode>;

// eslint-disable-next-line max-len
export function useWorkOrderTemplateNodes(): $ReadOnlyArray<WorkOrderTemplateNode> {
  const response = useLazyLoadQuery<WorkOrderTemplateNodesQuery>(
    workOrderTemplateNodesQuery,
  );
  const workOrderTemplatesData = response.workOrderTypes?.edges || [];
  const workOrderTemplates = workOrderTemplatesData
    .map(p => p.node)
    .filter(Boolean);
  return workOrderTemplates;
}

File: cli/pyworkforce/api/work_order.py
#!/usr/bin/env python3
# Copyright (c) 2004-present Facebook All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

from pyinventory.graphql.input.add_work_order_type import AddWorkOrderTypeInput
from pyinventory.graphql.mutation.add_work_order_type import AddWorkOrderTypeMutation
from pysymphony import SymphonyClient
from pysymphony.common.data_class import WorkOrder, WorkOrderType

from ..graphql.input.add_work_order import AddWorkOrderInput
from ..graphql.mutation.add_work_order import AddWorkOrderMutation


def add_work_order_type(client: SymphonyClient, name: str) -> WorkOrderType:
    """This function creates work order type with the given name

        Args:
            name (str): work order type name

        Returns:
            `pyworkforce.common.data_class.WorkOrderType`

        Example:
            ```
            client.add_work_order_type("Deployment work order")
            ```
    """
    result = AddWorkOrderTypeMutation.execute(
        client, AddWorkOrderTypeInput(name=name, checkListCategories=[])
    )
    return WorkOrderType(id=result.id)


def add_work_order(
    client: SymphonyClient, name: str, work_order_type: WorkOrderType
) -> WorkOrder:
    """This function creates work order of with the given name and type

        Args:
            name (str): work order name
            work_order_type (`pyworkforce.common.data_class.WorkOrderType`): work order type

        Returns:
            `pyworkforce.common.data_class.WorkOrder`

        Example:
            ```
            work_order_type = client.add_work_order_type("Deployment work order")
            client.add_work_order_type("new work order", work_order_type)
            ```
    """
    result = AddWorkOrderMutation.execute(
        client,
        AddWorkOrderInput(
            name=name,
            workOrderTypeId=work_order_type.id,
            properties=[],
            checkList=[],
            checkListCategories=[],
        ),
    )
    return WorkOrder(id=result.id, name=result.name)
